### ISSUE TITLE
FR: support for orgmode plain text files

### ISSUE THREAD

**Issue by Ypot** [koreader/koreader#10975]
**Does your feature request involve difficulty completing a task? Please describe.**
To read org files, the extension must be changed from .org to .txt

**Describe the solution you'd like**
As a first step, I would like KOReader to recognize .org as plain text files 

**Describe alternatives you've considered**
As medium term objective, it would be great to parse org syntax. There are already several parsers.
https://github.com/amake/org_parser
https://orgmode.org/worg/org-tools/
https://alphapapa.github.io/org-almanac/#Parsing

**Additional context**
For the GNU community, I think there is not a full feature ebook reader for org files.
 Just some apps like
https://github.com/orgzly
https://github.com/amake/orgro
Etc

**Comment by hius07** [comment]
As a first step, long-press on an .org file, Open with..., check Always use this engine for the type.

**Comment by pazos** [comment]
> As a first step, I would like KOReader to recognize .org as plain text files

It can be done progamatically, but I doubt it is a good option given how messy can be an org file that's displayed as plain text. IMHO better to stick with @hius07 suggestion, so power users can indulge self-damage doing such tasks ;)

> As medium term objective, it would be great to parse org syntax. There are already several parsers.

Parsing org syntax isn't the hardest part. The hardest part is what Emacs calls `fontification`. It means render the document based on a dynamic set of rules the user might override.

Not to mention dynamic folding, the entire `org-babel` code blocks and the miriad of org specifics that are too out of scope here.

> For the GNU community, I think there is not a full feature ebook reader for org files.
Just some great apps, but with different objectives, like:

> https://github.com/orgzly
> https://github.com/amake/orgro

Yup, they're great. Stick with them.

Also Org makes sense as a live document. If you want to have a snapshot of it, just to read it, you can use one of the many ways of exporting org to one of the formats well supported here.

Then, as an Emacs user, you can setup a hook that exports to desired format and push the format to your ereader, after saving such documents.

tl;dr: General integration with 3rd party tools is always welcome. GNU Emacs is not an exception. But we cannot attempt to support org files until a C library without external dependencies provides both a parser and a render for such documents. And even then we might consider it out-of-scope. Please use the right tool for the job :)

**Comment by Frenzie** [comment]
Also see the top search result: https://emacs.stackexchange.com/a/55655

**Comment by ryanwwest** [comment]
This, and maybe even more, native Markdown rendering would be nice and useful to many, but not enough for me to want to implement it so ¯\_(ツ)_/¯.

**Comment by pazos** [comment]
> This, and maybe even more, native Markdown rendering would be nice and useful to many, but not enough for me to want to implement it so ¯_(ツ)_/¯.

There're no such thing as a native markdown render. Markdown needs to be translated to html to be rendered as such. Or do you mean native support for mupdf to translate markdown to html on the fly?

Anyhow, markdown shouldn't be a major issue if somebody is interested. Mainly because it is a simple markup language that can be translated easily to html. We already bundle a pure lua markdown parser. If the requirement involves huge documents one can use already available quality libraries, like https://github.com/mity/md4c.

Sadly org mode is several orders of magnitude harder, because there's no correlation between an org document and html, other than the rules the document owner sets on emacs when exporting. Also, unlike markdown, "converting" an org document to html is a lossy proccess, while one can translate between markdown and html without too much trouble.

tl;dr: markdown is doable, org is not.

**Comment by ryanwwest** [comment]
> Or do you mean native support for mupdf to translate markdown to html on the fly?

Yes, I mean this. Older threads suggested exporting Markdown to HTML which works but I'd wager many don't want an extra HTML file lying around thereafter. 

> Anyhow, markdown shouldn't be a major issue if somebody is interested

Glad to know this is the case. There are many similar flavors of Markdown, so good to know that we could piggyback off one of them. 

I agree that org is significantly more difficult—I wonder if there's a partial rendering that could be done just for some of the most common features like hide/expand heading bullets, render code blocks/quotes, etc. But it's arguable whether a half-job is even helpful—seems like if there were a library that did a good job as a viewer we could use it but I don't know of any.

**Comment by Frenzie** [comment]
I already included a Markdown to HTML converter in KOReader years ago btw.

**Comment by pazos** [comment]
An "image" worth more than a thousand words.

```
cloc /usr/src/emacs/lisp/org/
     128 text files.
     127 unique files.
     127 files ignored.

github.com/AlDanial/cloc v 1.98  T=0.22 s (568.2 files/s, 639175.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Lisp                           127          14621          19110         109137
-------------------------------------------------------------------------------
SUM:                           127          14621          19110         109137
-------------------------------------------------------------------------------
```

```
cloc Software/koreader/frontend/ Software/koreader/plugins/
     462 text files.
     458 unique files.
       3 files ignored.

github.com/AlDanial/cloc v 1.98  T=1.02 s (448.8 files/s, 157525.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Lua                            450          11872          18660         128147
Bourne Shell                     5            166             92           1090
JSON                             2              0              0            694
Markdown                         1              7              0             19
-------------------------------------------------------------------------------
SUM:                           458          12045          18752         129950
-------------------------------------------------------------------------------
```

Also, org itself is powerful within emacs as it inherits most of the core emacs capabilities. For instance: syntax highlighting from relevant major modes, cycling visibility or narrow to subtree/block from common narrow capabilities, etc.

> I agree that org is significantly more difficult—I wonder if there's a partial rendering that could be done just for some of the most common features like hide/expand heading bullets, render code blocks/quotes

The problem I see is that we need to parse the entire document only to be able to render a minimal subset of it. Even the android apps available struggle to have some of the basic features working (ie: syntax highlighting) and there is possible to rely on javascript on a webview to do those kind of tasks.

I'm afraid you would need a rich environment, such as a modern web browser or a mobile app to be able to do 10% of what you can do on emacs.

https://www.filestash.app/ is a great self hosted service that's able to manage org documents. Makes sense to use those apps as both a regular computer & browser or a phone/tablet are better writting environments than KOReader on eink devices.

**Comment by Ypot** [comment]
> > 
> 
> I agree that org is significantly more difficult—I wonder if there's a partial rendering that could be done just for some of the most common features like hide/expand heading bullets, render code blocks/quotes, etc. 

For just reading books, maybe it would be enough to see the fortification of orgmode: 
_underline_
*Bold*
Etc.

**Comment by pazos** [comment]
> > > 
> > 
> > 
> > I agree that org is significantly more difficult—I wonder if there's a partial rendering that could be done just for some of the most common features like hide/expand heading bullets, render code blocks/quotes, etc.
> 
> For just reading books, maybe it would be enough to see the fortification of orgmode: _underline_ _Bold_ Etc.

Just use `ox-html` and get a nice html document. Or pandoc and turn the document into a beautiful epub :)

**Comment by Frenzie** [comment]
You could also install emacs and run that ox-html thing from a script.

**Comment by ryanwwest** [comment]
Aside, but Emacs itself is officially being ported to Android with release 30.1: https://www.emacswiki.org/emacs/EmacsOnAndroid. Maybe not mobile-friendly enough to start but that would give the full power of reading/writing orgmode.
